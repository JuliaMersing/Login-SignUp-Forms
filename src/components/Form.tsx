import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { useToast } from 'react-native-toast-notifications';
import { newUser } from '../redux/registerSlice';
import { verifyEmail, verifyPassword } from '../utils/validations';
import Input from './Input';
import Checkbox from './Checkbox';
import Header from './Header';
import Button from './Button';

type FormProps = {
  isSignUp?: boolean;
};

const Form: React.FunctionComponent<FormProps> = ({ isSignUp }: FormProps) => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [confirmedPassword, setConfirmedPassword] = useState('');
  const [remember, setRemember] = useState(true);
  const [emailError, setEmailError] = useState('');
  const [passwordError, setPasswordError] = useState('');
  const [confirmedPasswordError, setConfirmedPasswordError] = useState('');
  const validLogin = passwordError === '' && emailError === '' && confirmedPasswordError === '';

  const dispatch = useDispatch();

  const toast = useToast();

  const handleEmail = (event: any): void => {
    setEmail(event.target.value);
  };

  const handlePassword = (event: any): void => {
    setPassword(event.target.value);
  };

  const handleConfirmedPassword = (event: any): void => {
    setConfirmedPassword(event.target.value);
  };

  const handleRememberChange = (event: any): void => {
    const { checked } = event.target;
    setRemember(checked);
  };

  const handleEmailOnBlur = (event: any): void => {
    const error = verifyEmail(event.target.value);
    setEmailError(error);
  };

  const handlePasswordOnBlur = (event: any): void => {
    const error = verifyPassword(event.target.value);
    setPasswordError(error);
  };

  const handleConfirmedPasswordOnBlur = (event: any): void => {
    const error = event.current.target;
    setConfirmedPasswordError(error);
  };

  const onFormSubmit = (event: any): void => {
    event.preventDefault();
    if (!validLogin) {
      toast.show('Form has errors', {
        type: 'danger',
        placement: 'top',
      });
      return;
    }
    dispatch(
      newUser({
        email,
        password,
      }),
    );
  };

  return (
    <div className='container-form'>
      <div className='container-header'>
        {isSignUp ? (
          <Header
            heading='Sign Up to create an account'
            paragraph='Already have an account?'
            href='/'
            linkParagraph='Login'
          />
        ) : (
          <Header heading='Login' paragraph="Don't have an account?" href='/signUp' linkParagraph='SignUp' />
        )}

        <form className='mt-8 space-y-6' action='#' method='POST' onSubmit={onFormSubmit}>
          <input type='hidden' name='remember' value='true' />
          <div className='rounded-md shadow-sm -space-y-px'>
            <Input
              id='email'
              type='email'
              value={email}
              placeholder='Email'
              onChange={handleEmail}
              onBlur={handleEmailOnBlur}
              error={emailError}
              className={passwordError ? 'input-error' : 'input'}
            />
            <Input
              id='password'
              type='password'
              value={password}
              placeholder='Password'
              onChange={handlePassword}
              onBlur={handlePasswordOnBlur}
              error={passwordError}
              className={passwordError ? 'input-error' : 'input'}
            />
            {isSignUp && (
              <Input
                id='password'
                type='password'
                value={confirmedPassword}
                placeholder='Confirm password'
                onChange={handleConfirmedPassword}
                onBlur={handleConfirmedPasswordOnBlur}
                error={confirmedPasswordError}
                className={passwordError ? 'input-error' : 'input'}
              />
            )}
          </div>
          <div className='container-checkbox'>
            <Checkbox onChange={handleRememberChange} check={remember} />
            <div className='text-sm'>
              <a href='#' className='container-footer'>
                {' '}
                Forgot your password?{' '}
              </a>
            </div>
          </div>
          <Button> Sign in </Button>
        </form>
      </div>
    </div>
  );
};

export default Form;
