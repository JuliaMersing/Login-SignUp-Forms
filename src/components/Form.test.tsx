import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Form from './Form';

const mockUseDispatch = jest.fn();
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockUseDispatch,
}));

describe('Form', () => {
  const setUp = () => {
    const utils = render(<Form />);

    const formButton = screen.getByRole('button');
    const rememberCheck = screen.getByRole('checkbox');

    return {
      ...utils,
      formButton,
      rememberCheck,
    };
  };
  test('should render component and match snapshot', () => {
    const { container } = render(<Form />);
    expect(container).toMatchSnapshot();
  });
  test('should send form and call useDispatch when button is clicked', async () => {
    const {
      formButton,
    } = setUp();

    await userEvent.click(formButton);
    expect(mockUseDispatch).toHaveBeenCalled();
  });
});
