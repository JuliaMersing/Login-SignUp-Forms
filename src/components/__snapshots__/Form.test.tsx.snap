// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render component and match snapshot 1`] = `
<div>
  <div
    class="container-form"
  >
    <div
      class="container-header"
    >
      <div
        class="mb-10"
      >
        <div
          class="flex justify-center"
        >
          <i
            class=" fa-solid fa-mug-hot icon"
          />
        </div>
        <h2
          class="heading"
        >
          Login
        </h2>
        <p
          class="paragraphs-header"
        >
          Don't have an account?
          <a
            class="link"
            href="/signUp"
          >
            SignUp
          </a>
        </p>
      </div>
      <form
        action="#"
        class="mt-8 space-y-6"
        method="POST"
      >
        <input
          name="remember"
          type="hidden"
          value="true"
        />
        <div
          class="rounded-md shadow-sm -space-y-px"
        >
          <div
            class="mb-6"
          >
            <input
              class="input"
              id="email"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            class="mb-6"
          >
            <input
              class="input"
              id="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
        </div>
        <div
          class="container-checkbox"
        >
          <div
            class="flex items-center"
          >
            <input
              checked=""
              class="checkbox-input"
              id="remember-me"
              name="remember-me"
              type="checkbox"
            />
            <label
              class="paragraphs"
              for="remember-me"
            >
               Remember me 
            </label>
          </div>
          <div
            class="text-sm"
          >
            <a
              class="container-footer"
              href="#"
            >
               Forgot your password? 
            </a>
          </div>
        </div>
        <div>
          <button
            class="button-container"
            type="submit"
          >
            <span
              class="span"
            >
              <svg
                aria-hidden="true"
                class="button-lock"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M5 9V7a5 5 0 0110 0v2a2 2 0 012 2v5a2 2 0 01-2 2H5a2 2 0 01-2-2v-5a2 2 0 012-2zm8-2v2H7V7a3 3 0 016 0z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
             Sign in 
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
